### PR TITLE
Adding install/uninstall flags to daemon

### PR DESCRIPTION
**Pull request by PoppySeedPlehzr**

### CODE REVIEW

File: osquery/main/windows/daemon.cpp [hunk 1]
```diff
@@ -54,7 +66,142 @@ void DebugPrintf(const char *fmt, ...) {
...
+    // Wait 3 seconds to give the service an opportunity to stop.
+    Sleep(3000);
+    QueryServiceStatus(schService, &ssSvcStatus);
+    if (ssSvcStatus.dwCurrentState != SERVICE_STOPPED) {
+      return Status(1);
+    }
+  }
+
+  // Now remove the service by calling DeleteService.
+  if (!DeleteService(schService)) {
```
Comment (theopolis) [inline osquery/main/windows/daemon.cpp]: To save some code, can you:
```cpp
auto s = DeleteService(schService);
CloseServiceHandle(schSCManager);
CloseServiceHandle(schService);
return Status(s ? 0 : 1);
```

File: osquery/main/windows/daemon.cpp [hunk 2]
```diff
@@ -54,7 +66,142 @@ void DebugPrintf(const char *fmt, ...) {
...
+  }
+
+  SERVICE_STATUS ssSvcStatus = {};
+  if (ssStatus.dwCurrentState != SERVICE_STOPPED) {
+    ControlService(schService, SERVICE_CONTROL_STOP, &ssSvcStatus);
+    // Wait 3 seconds to give the service an opportunity to stop.
+    Sleep(3000);
+    QueryServiceStatus(schService, &ssSvcStatus);
+    if (ssSvcStatus.dwCurrentState != SERVICE_STOPPED) {
+      return Status(1);
```
Comment (theopolis) [inline osquery/main/windows/daemon.cpp]: Do you need to close some handles here?

File: osquery/main/windows/daemon.cpp [hunk 3]
```diff
@@ -54,7 +66,142 @@ void DebugPrintf(const char *fmt, ...) {
...
+                             nullptr,
+                             nullptr,
+                             nullptr, // User Account. nullptr => LOCAL SYSTEM
+                             nullptr);
+
+  if (schService == nullptr) {
+    return Status(1);
+  }
+
+  if (schSCManager) {
```
Comment (theopolis) [inline osquery/main/windows/daemon.cpp]: `if` what?

File: osquery/main/windows/daemon.cpp [hunk 4]
```diff
@@ -54,7 +66,142 @@ void DebugPrintf(const char *fmt, ...) {
...
+    CloseServiceHandle(schSCManager);
+    schSCManager = nullptr;
+  }
+  if (schService) {
```
Comment (theopolis) [inline osquery/main/windows/daemon.cpp]: Didn't you check against `== nullptr` above?

**Comment by osqueryer** [comment]
:no_good: The commit 64a8576ca33581213fa9ec746fc0211cbea5d384 (Job results: [60](https://jenkins.osquery.io/job/osqueryPullRequestAudit/60/)) failed the code audit.

File: osquery/main/windows/daemon.cpp [hunk 5]
```diff
@@ -54,7 +66,124 @@ void DebugPrintf(const char *fmt, ...) {
...
+
+  if (schSCManager == nullptr) {
+    return Status(1);
+  }
+
+  SC_HANDLE schService =
+      OpenService(schSCManager, kServiceName.c_str(), SERVICE_ALL_ACCESS);
+
+  if (schService != nullptr) {
+    return Status(1);
```
Comment (theopolis) [inline osquery/main/windows/daemon.cpp]: Need to close `schSCManager` here.

File: osquery/main/windows/daemon.cpp [hunk 6]
```diff
@@ -54,7 +66,124 @@ void DebugPrintf(const char *fmt, ...) {
   va_end(vl);
 }
 
-void UpdateServiceStatus(DWORD controls, DWORD state, DWORD exit_code, DWORD checkpoint) {
+/// Install osqueryd as a service given the path to the binary
+Status installOsqueryService(const std::string& binPath) {
```
Comment (theopolis) [inline osquery/main/windows/daemon.cpp]: You're implicitly creating a r-value `std::string` then later only accessing the `char*` internal representation. If you pass a `const char* const binPath` you'll save the conversion.

File: osquery/main/windows/daemon.cpp [hunk 7]
```diff
@@ -82,9 +211,26 @@ void WINAPI ServiceControlHandler(DWORD control_code) {
...
-void daemonEntry(int argc, char *argv[]) {
+void daemonEntry(int argc, char* argv[]) {
   osquery::Initializer runner(argc, argv, osquery::ToolType::DAEMON);
 
+  // Options for installing or uninstalling the osqueryd as a service
+  if (osquery::FLAGS_install) {
+    Status s = osquery::installOsqueryService(std::string(argv[0]));
+    if (!s.ok()) {
+      LOG(ERROR) << "Unable to install the osqueryd service.";
+      exit(0);
```
Comment (theopolis) [inline osquery/main/windows/daemon.cpp]: Just return and fall-through to the `main`'s `return 0`.

File: osquery/main/windows/daemon.cpp [hunk 8]
```diff
@@ -82,9 +211,26 @@ void WINAPI ServiceControlHandler(DWORD control_code) {
...
+    }
+    osquery::startOsqueryService();
+    exit(0);
+  } else if (osquery::FLAGS_uninstall) {
+    Status s = osquery::uninstallOsqueryService();
+    if (!s.ok()) {
+      LOG(ERROR) << "Unable to uninstall the osqueryd service.";
+    }
+    exit(0);
```
Comment (theopolis) [inline osquery/main/windows/daemon.cpp]: Same as above.

File: osquery/main/windows/daemon.cpp [hunk 9]
```diff
@@ -82,9 +211,26 @@ void WINAPI ServiceControlHandler(DWORD control_code) {
...
 }
 
-void daemonEntry(int argc, char *argv[]) {
+void daemonEntry(int argc, char* argv[]) {
   osquery::Initializer runner(argc, argv, osquery::ToolType::DAEMON);
 
+  // Options for installing or uninstalling the osqueryd as a service
+  if (osquery::FLAGS_install) {
+    Status s = osquery::installOsqueryService(std::string(argv[0]));
+    if (!s.ok()) {
```
Comment (theopolis) [inline osquery/main/windows/daemon.cpp]: If you're not doing anything with the `Status` code/message then `if(osquery::installOsqueryService) {` saves code.

File: osquery/main/windows/daemon.cpp [hunk 10]
```diff
@@ -82,9 +211,26 @@ void WINAPI ServiceControlHandler(DWORD control_code) {
...
+      LOG(ERROR) << "Unable to install the osqueryd service.";
+      exit(0);
+    }
+    osquery::startOsqueryService();
+    exit(0);
+  } else if (osquery::FLAGS_uninstall) {
+    Status s = osquery::uninstallOsqueryService();
+    if (!s.ok()) {
```
Comment (theopolis) [inline osquery/main/windows/daemon.cpp]: Same as above.

File: osquery/main/windows/daemon.cpp [hunk 11]
```diff
@@ -82,9 +211,26 @@ void WINAPI ServiceControlHandler(DWORD control_code) {
...
 
-void daemonEntry(int argc, char *argv[]) {
+void daemonEntry(int argc, char* argv[]) {
   osquery::Initializer runner(argc, argv, osquery::ToolType::DAEMON);
 
+  // Options for installing or uninstalling the osqueryd as a service
+  if (osquery::FLAGS_install) {
+    Status s = osquery::installOsqueryService(std::string(argv[0]));
+    if (!s.ok()) {
+      LOG(ERROR) << "Unable to install the osqueryd service.";
```
Comment (theopolis) [inline osquery/main/windows/daemon.cpp]: nit, typically we don't add "`.`"s into the log lines. 💃

File: osquery/main/windows/daemon.cpp (hunk 6 above)
Comment (theopolis) [inline osquery/main/windows/daemon.cpp]: Also `installOsqueryService` in the `osquery` namespace is a bit redundant, any reason not to prefer `installService`?

File: osquery/main/windows/daemon.cpp [hunk 12]
```diff
@@ -82,9 +211,26 @@ void WINAPI ServiceControlHandler(DWORD control_code) {
...
+      exit(0);
+    }
+    osquery::startOsqueryService();
+    exit(0);
```
Comment (theopolis) [inline osquery/main/windows/daemon.cpp]: same

**Comment by osqueryer** [comment]
:-1: The commit b18aa7b92fa3db4837bfc569cd34aae3c041d495 (Job results: [3085](https://jenkins.osquery.io/job/osqueryPullRequestBuild/3085/)) failed one or more tests (**OS X/Linux**).

**Pull request merged**